semantic/src/main/kotlin/tools/samt/semantic/Types.kt

```diff
@@ -0,0 +1,201 @@
+    }
+
+    class RequestResponseOperation(
+        name: String,
+        parameters: List<Parameter>,
+        var returnType: TypeReference?,
+        var raisesTypes: List<TypeReference>,
+        val isAsync: Boolean,
+    ) : Operation(name, parameters)
+
+    class OnewayOperation(
+        name: String,
+        parameters: List<Parameter>,
+    ) : Operation(name, parameters)
+
+    override val humanReadableName: String = name
+}
+
+data class ProviderType(
+    val name: String,
+    val implements: List<Implements>,
+    val transport: Transport,
+    override val definition: ProviderDeclarationNode,
+) : CompoundType, UserDefinedType {
+    data class Implements(
+        var service: TypeReference,
+        var operations: List<ServiceType.Operation>,
+    )
+
+    data class Transport(
+        val name: String,
+        val configuration: Any?,
+    )
+
+    override val humanReadableName: String = name
+}
+
+data class ConsumerType(
+    val provider: TypeReference,
+    var uses: List<Uses>,
+    override val definition: ConsumerDeclarationNode,
+) : CompoundType, UserDefinedType {
+    data class Uses(
+        val service: TypeReference,
+        var operations: List<ServiceType.Operation>,
+    )
+
+    override val humanReadableName: String = "consumer for ${provider.humanReadableName}"
+}
+
+interface TypeReference {
+    val humanReadableName: String
+}
+
+data class UnresolvedTypeReference(
+    val expression: ExpressionNode,
+) : TypeReference {
+    override val humanReadableName: String
+        get() = "<Unresolved Type at ${expression.location}>"
+}
+
+data class ResolvedTypeReference(
+    val isOptional: Boolean,
+    var constraints: List<Constraint>,
+    val type: Type,
+) : TypeReference {
+    override val humanReadableName: String
+        get() = buildString {
+            append(type.humanReadableName)
+
+            if (isOptional) {
+                append('?')
+            }
+
+            if (constraints.isNotEmpty()) {
+                append(' ')
+                append('(')
+                constraints.joinTo(this, ", ") { it.humanReadableName }
+                append(')')
+            }
+        }
+
+    sealed interface Constraint {
+        val humanReadableName: String
+
+        data class Range(
+            val lowerBound: Number?,
+            val upperBound: Number?,
+        ) : Constraint {
+            override val humanReadableName: String
+                get() = "range(${lowerBound ?: '*'}..${upperBound ?: '*'})"
+        }
+
+        data class Size(
+            val lowerBound: Long?,
+            val upperBound: Long?,
+        ) : Constraint {
+            override val humanReadableName: String
+                get() = "size(${lowerBound ?: '*'}..${upperBound ?: '*'})"
+        }
+
+        data class Pattern(
+            val pattern: String,
+        ) : Constraint {
+            override val humanReadableName: String
+                get() = "pattern($pattern)"
+        }
```

Comment:
This is missing the explicit value constraint we've declared in FR `FR-LF-TYPECONSTRAINTS` (`Value limit string, number or boolean to specific value`)

Reply:
I hadn't looked into the specific constraints, but good catch I added a `value(...)` constraint so it won't be forgotten later